includes/Speculative_Prerendering.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Class Speculative_Prerendering
+ *
+ * @link      https://github.com/googleforcreators/web-stories-wp
+ *
+ * @copyright 2024 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ */
+
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare(strict_types = 1);
+
+namespace Google\Web_Stories\Admin;
+
+use Google\Web_Stories\Service_Base;
+
+/**
+ * Speculative_Prerendering class.
+ */
+class Speculative_Prerendering extends Service_Base {
+
+	/**
+	 * Array defining the matches for different pages.
+	 *
+	 * @var array<string,array<string>>
+	 */
+	private array $matches = [
+		'dashboard'   => [ 'post.php?post=*&action=edit', 'post-new.php?post_type=web-story' ],
+		'all_stories' => [ 'post.php?post=*&action=edit', 'post-new.php?post_type=web-story', '/web-stories*' ],
+		'archive'     => [ '/web-stories*' ],
+	];
+
+	/**
+	 * Runs on instantiation.
+	 */
+	public function register(): void {
+		add_action( 'admin_enqueue_scripts', [ $this, 'load_rules' ] );
+	}
+
+	/**
+	 * Loads the prerendering rules based on the current page.
+	 *
+	 * @param string $hook The current page hook.
+	 */
+	public function load_rules( string $hook ): void {
+		$rules = [];
+		if ( 'web-story_page_stories-dashboard' === $hook ) {
```

Comment:
Instead of hardcoding the hook suffix like this, we should inject the `Dashboard` service and get it from `$this->dashboard->get_hook_suffix( 'stories-dashboard' )`

Reply:
`$this->dashboard->get_hook_suffix( 'stories-dashboard' )` is always returning `false` for me. 

```php
public function get_hook_suffix( string $key ) {
  return $this->hook_suffix[ $key ] ?? false;
}
```
Upon taking a closer look, `$this->hook_suffix[ $key ]` gets us `web-story_page_stories-dashboard`, but the function returns `false`

Reply:
The doc string has - 

```
@return bool|string The dashboard page's hook_suffix, or false if the user does not have the capability required.
```

Maybe this has something to do with user capabilities.

Reply:
You'll need to add `Dashboard::class` to `\Google\Web_Stories\Plugin::get_shared_instances()` in order to address this. Otherwise the `Dashboard` instance you get is a new one which at this point won't have registered the menu item yet. Think of this as turning `Dashboard` into a singleton.